parser/prism/Translator.h

```diff
@@ -23,16 +23,20 @@ class Translator final {
     const absl::Span<const ParseError> parseErrors;
 
     // Whether to directly desugar during in the Translator, or wait until the usual `Desugar.cc` code path.
-    bool directlyDesugar;
+    const bool directlyDesugar;
+
+    // Unique counters used to create synthetic names via `ctx.state.freshNameUnique`.
+    // - The storage integers either store an "active" count used by a translator and some of its children,
+    //   or a dummy value (int min).
+    // - The pointer variables point to the "active" count for each translator,
+    //   which is either pointing to its own storage, or to a parent's storage.
+    uint16_t parserUniqueCounterStorage;  // Minics the `Builder::Impl.uniqueCounter_` in `parser/Builder.cc`
+    uint32_t desugarUniqueCounterStorage; // Minics the `DesugarContext.uniqueCounter`  in `ast/desugar/Desugar.cc`
+    uint16_t *const parserUniqueCounter;  // Points to the active `parserUniqueCounterStorage`
+    uint32_t *const desugarUniqueCounter; // Points to the active `desugarUniqueCounterStorage`
```

Comment:
Our unique names wouldn't necessarily have to match the non-Prism pipeline, but it would be tricky to make the tests ignore the differences while still ensuring correctness.

We'll match the old pipeline exactly for now, but once all is said and done, we can fuse these into a single counter.

Also notably, `parser/Builder.cc` never resets its counter the way Desugar does:

https://github.com/sorbet/sorbet/blob/55f3f4a13fe43a90f03b14e0be72d4344324cdae/ast/desugar/Desugar.cc#L365-L368

In the future when we merge these, the synthetic names for block args and keyword rest parameters can be shortened.

Reply:
Just to confirm, these are never reused across threads, right? If that's the case, I'd suggest making them into atomic values instead.

Given that they're always shared between parent and child `Translator` values, it might be easier to make them into `std::shared_ptr<int>` values instead. That way you wouldn't need to manage the pointers at all, and would get the same sharing implicitly when you copy.